File: arduinobot_task_planner/arduinobot_task_planner/robot_interface_test.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from control_msgs.action import FollowJointTrajectory
from trajectory_msgs.msg import JointTrajectoryPoint
from builtin_interfaces.msg import Duration
import threading
import time


class RobotInterface(Node):
    def __init__(self):
        super().__init__('robot_interface')

        self.arm_client = ActionClient(
            self,
            FollowJointTrajectory,
            '/arm_controller/follow_joint_trajectory'
        )

        self.gripper_client = ActionClient(
            self,
            FollowJointTrajectory,
            '/gripper_controller/follow_joint_trajectory'
        )

        self.arm_presets = {
            'home':  [0.0,  0.0,   0.0],
            'pick':  [0.8,  0.5,  -0.5],
            'place': [-0.8, 0.5,  -0.5],
        }

        self.GRIPPER_OPEN   = 0.0
        self.GRIPPER_CLOSED = -1.2

        self.is_executing = False
        self._pending_duration = 1.0

        self.get_logger().info('Waiting for action servers...')
        self.arm_client.wait_for_server()
        self.gripper_client.wait_for_server()
        self.get_logger().info('Action servers ready')

        self.print_help()

        self.input_thread = threading.Thread(
            target=self.input_loop,
            daemon=True
        )
        self.input_thread.start()

    def print_help(self):
        print('\n============================')
        print('   Arduinobot Interface')
        print('============================')
        print('  home   — home position')
        print('  pick   — pick position')
        print('  place  — place position')
        print('  open   — open gripper')
        print('  close  — close gripper')
        print('  task   — pick and place sequence')
        print('  help   — show this menu')
        print('  quit   — exit')
        print('============================\n')

    def send_arm(self, positions, duration_sec=2):
        if self.is_executing:
            print('Robot moving — wait...')
            return

        goal = FollowJointTrajectory.Goal()
        goal.trajectory.joint_names = ['joint_1', 'joint_2', 'joint_3']
        point = JointTrajectoryPoint()
        point.positions = positions
        point.time_from_start = Duration(sec=duration_sec)
        goal.trajectory.points = [point]

        self._pending_duration = float(duration_sec)
        self.is_executing = True
        future = self.arm_client.send_goal_async(goal)
        future.add_done_callback(self.goal_response_callback)
        self.get_logger().info(f'Arm → {positions}')

    def send_gripper(self, position, duration_sec=2):
        if self.is_executing:
            print('Robot moving — wait...')
            return

        goal = FollowJointTrajectory.Goal()
        goal.trajectory.joint_names = ['joint_4']
        point = JointTrajectoryPoint()
        point.positions = [position]
        point.time_from_start = Duration(sec=duration_sec)
        goal.trajectory.points = [point]

        self._pending_duration = float(duration_sec)
        self.is_executing = True
        future = self.gripper_client.send_goal_async(goal)
        future.add_done_callback(self.goal_response_callback)
        state = 'OPEN' if position >= 0.0 else 'CLOSED'
        self.get_logger().info(f'Gripper → {state}')

    def goal_response_callback(self, future):
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().error('Goal rejected')
            self.is_executing = False
            return
        self.get_logger().info('Goal accepted')
        result_future = goal_handle.get_result_async()
        result_future.add_done_callback(self.result_callback)

    def result_callback(self, future):
        # Attendi la durata effettiva del movimento prima di sbloccare
        time.sleep(self._pending_duration)
        self.is_executing = False
        self.get_logger().info('Movement completed')

    def execute_task(self):
        """
        Sequenza corretta:
        1. Vai in posizione pick
        2. Apri gripper
        3. Chiudi gripper
        4. Vai in posizione place
        5. Apri gripper
        6. Chiudi gripper
        7. Torna in posizione home
        """
        self.get_logger().info('Starting pick and place sequence...')

        steps = [
            ('arm',     self.arm_presets['pick'],  2),
            ('gripper', self.GRIPPER_OPEN,          2),
            ('gripper', self.GRIPPER_CLOSED,        2),
            ('arm',     self.arm_presets['place'],  2),
            ('gripper', self.GRIPPER_OPEN,          2),
            ('gripper', self.GRIPPER_CLOSED,        2),
            ('arm',     self.arm_presets['home'],   2),
        ]

        for controller, target, duration in steps:
            while self.is_executing:
                time.sleep(0.1)

            if controller == 'arm':
                self.send_arm(target, duration)
            else:
                self.send_gripper(target, duration)

            time.sleep(duration + 0.5)

        while self.is_executing:
            time.sleep(0.1)

        self.get_logger().info('Pick and place sequence completed')

    def input_loop(self):
        while True:
            try:
                cmd = input('\n> ').strip().lower()
            except EOFError:
                break

            if cmd in self.arm_presets:
                self.send_arm(self.arm_presets[cmd])

            elif cmd == 'open':
                self.send_gripper(self.GRIPPER_OPEN)

            elif cmd == 'close':
                self.send_gripper(self.GRIPPER_CLOSED)

            elif cmd == 'task':
                if self.is_executing:
                    print('Robot moving — wait...')
                else:
                    t = threading.Thread(
                        target=self.execute_task,
                        daemon=True
                    )
                    t.start()

            elif cmd == 'help':
                self.print_help()

            elif cmd == 'quit':
                self.get_logger().info('Exiting program')
                break

            else:
                print(f'Unknown command: "{cmd}"')
                print('Type "help" for available commands')


def main(args=None):
    rclpy.init(args=args)
    node = RobotInterface()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: arduinobot_task_planner_cpp/src/robot_interface_test.cpp
#include <memory>
#include <vector>
#include <string>
#include <thread>
#include <atomic>
#include <chrono>
#include <iostream>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

#include "control_msgs/action/follow_joint_trajectory.hpp"
#include "trajectory_msgs/msg/joint_trajectory_point.hpp"
#include "builtin_interfaces/msg/duration.hpp"

using FollowJointTrajectory = control_msgs::action::FollowJointTrajectory;

class RobotInterface : public rclcpp::Node
{
public:
    RobotInterface()
        : Node("robot_interface"),
          is_executing_(false),
          pending_duration_(1.0)
    {
        // Action clients
        arm_client_ = rclcpp_action::create_client<FollowJointTrajectory>(
            this,
            "/arm_controller/follow_joint_trajectory");

        gripper_client_ = rclcpp_action::create_client<FollowJointTrajectory>(
            this,
            "/gripper_controller/follow_joint_trajectory");

        // Presets
        arm_presets_["home"]  = {0.0,  0.0,  0.0};
        arm_presets_["pick"]  = {0.8,  0.5, -0.5};
        arm_presets_["place"] = {-0.8, 0.5, -0.5};

        GRIPPER_OPEN_ = 0.0;
        GRIPPER_CLOSED_ = -1.2;

        RCLCPP_INFO(this->get_logger(), "Waiting for action servers...");

        if (!arm_client_->wait_for_action_server(std::chrono::seconds(5)) ||
            !gripper_client_->wait_for_action_server(std::chrono::seconds(5)))
        {
            RCLCPP_WARN(this->get_logger(), "Action servers not ready yet");
        }

        RCLCPP_INFO(this->get_logger(), "Action servers ready");

        printHelp();

        input_thread_ = std::thread(&RobotInterface::inputLoop, this);
        input_thread_.detach();
    }

private:

    // =========================
    // ACTION SEND ARM
    // =========================
    void sendArm(const std::vector<double> &positions, double duration_sec = 2.0)
    {
        if (is_executing_)
        {
            std::cout << "Robot moving — wait...\n";
            return;
        }

        FollowJointTrajectory::Goal goal;
        goal.trajectory.joint_names = {"joint_1", "joint_2", "joint_3"};

        trajectory_msgs::msg::JointTrajectoryPoint point;
        point.positions = positions;

        builtin_interfaces::msg::Duration dur;
        dur.sec = static_cast<int32_t>(duration_sec);
        point.time_from_start = dur;

        goal.trajectory.points.push_back(point);

        pending_duration_ = duration_sec;
        is_executing_ = true;

        RCLCPP_INFO(this->get_logger(), "Arm command sent");

        auto send_goal_options =
            rclcpp_action::Client<FollowJointTrajectory>::SendGoalOptions();

        send_goal_options.goal_response_callback =
            std::bind(&RobotInterface::goalResponseCallback, this, std::placeholders::_1);

        send_goal_options.result_callback =
            std::bind(&RobotInterface::resultCallback, this, std::placeholders::_1);

        arm_client_->async_send_goal(goal, send_goal_options);
    }

    // =========================
    // ACTION SEND GRIPPER
    // =========================
    void sendGripper(double position, double duration_sec = 2.0)
    {
        if (is_executing_)
        {
            std::cout << "Robot moving — wait...\n";
            return;
        }

        FollowJointTrajectory::Goal goal;
        goal.trajectory.joint_names = {"joint_4"};

        trajectory_msgs::msg::JointTrajectoryPoint point;
        point.positions = {position};

        builtin_interfaces::msg::Duration dur;
        dur.sec = static_cast<int32_t>(duration_sec);
        point.time_from_start = dur;

        goal.trajectory.points.push_back(point);

        pending_duration_ = duration_sec;
        is_executing_ = true;

        std::string state = (position >= 0.0) ? "OPEN" : "CLOSED";
        RCLCPP_INFO(this->get_logger(), "Gripper → %s", state.c_str());

        auto send_goal_options =
            rclcpp_action::Client<FollowJointTrajectory>::SendGoalOptions();

        send_goal_options.goal_response_callback =
            std::bind(&RobotInterface::goalResponseCallback, this, std::placeholders::_1);

        send_goal_options.result_callback =
            std::bind(&RobotInterface::resultCallback, this, std::placeholders::_1);

        gripper_client_->async_send_goal(goal, send_goal_options);
    }

    // =========================
    // CALLBACKS
    // =========================
    void goalResponseCallback(
        rclcpp_action::ClientGoalHandle<FollowJointTrajectory>::SharedPtr goal_handle)
    {
        if (!goal_handle)
        {
            RCLCPP_ERROR(this->get_logger(), "Goal rejected");
            is_executing_ = false;
            return;
        }

        RCLCPP_INFO(this->get_logger(), "Goal accepted");
    }

    void resultCallback(
        const rclcpp_action::ClientGoalHandle<FollowJointTrajectory>::WrappedResult &)
    {
        std::this_thread::sleep_for(
            std::chrono::milliseconds(static_cast<int>(pending_duration_ * 1000)));

        is_executing_ = false;
        RCLCPP_INFO(this->get_logger(), "Movement completed");
    }

    // =========================
    // TASK SEQUENCE
    // =========================
    void executeTask()
    {
        RCLCPP_INFO(this->get_logger(), "Starting pick and place sequence...");

        struct Step
        {
            std::string type;
            std::vector<double> arm_target;
            double gripper_target;
            bool is_arm;
            double duration;
        };

        std::vector<Step> steps = {
            {"arm",     arm_presets_["pick"],  0.0, true,  2},
            {"gripper", {}, GRIPPER_OPEN_,     false, 2},
            {"gripper", {}, GRIPPER_CLOSED_,   false, 2},
            {"arm",     arm_presets_["place"], 0.0, true,  2},
            {"gripper", {}, GRIPPER_OPEN_,     false, 2},
            {"gripper", {}, GRIPPER_CLOSED_,   false, 2},
            {"arm",     arm_presets_["home"],  0.0, true,  2},
        };

        for (auto &s : steps)
        {
            while (is_executing_)
                std::this_thread::sleep_for(std::chrono::milliseconds(100));

            if (s.is_arm)
                sendArm(s.arm_target, s.duration);
            else
                sendGripper(s.gripper_target, s.duration);

            std::this_thread::sleep_for(
                std::chrono::milliseconds(static_cast<int>(s.duration * 1000 + 500)));
        }

        while (is_executing_)
            std::this_thread::sleep_for(std::chrono::milliseconds(100));

        RCLCPP_INFO(this->get_logger(), "Pick and place sequence completed");
    }

    // =========================
    // INPUT LOOP
    // =========================
    void inputLoop()
    {
        std::string cmd;

        while (rclcpp::ok())
        {
            std::cout << "\n> ";
            std::getline(std::cin, cmd);

            for (auto &c : cmd)
                c = std::tolower(c);

            if (arm_presets_.count(cmd))
            {
                sendArm(arm_presets_[cmd]);
            }
            else if (cmd == "open")
            {
                sendGripper(GRIPPER_OPEN_);
            }
            else if (cmd == "close")
            {
                sendGripper(GRIPPER_CLOSED_);
            }
            else if (cmd == "task")
            {
                if (is_executing_)
                {
                    std::cout << "Robot moving — wait...\n";
                }
                else
                {
                    std::thread(&RobotInterface::executeTask, this).detach();
                }
            }
            else if (cmd == "help")
            {
                printHelp();
            }
            else if (cmd == "quit")
            {
                RCLCPP_INFO(this->get_logger(), "Exiting program");
                rclcpp::shutdown();
                break;
            }
            else
            {
                std::cout << "Unknown command: " << cmd << "\n";
                std::cout << "Type 'help' for available commands\n";
            }
        }
    }

    // =========================
    // HELP
    // =========================
    void printHelp()
    {
        std::cout << "\n============================\n";
        std::cout << "   Arduinobot Interface\n";
        std::cout << "============================\n";
        std::cout << " home   — home position\n";
        std::cout << " pick   — pick position\n";
        std::cout << " place  — place position\n";
        std::cout << " open   — open gripper\n";
        std::cout << " close  — close gripper\n";
        std::cout << " task   — pick and place sequence\n";
        std::cout << " help   — show menu\n";
        std::cout << " quit   — exit\n";
        std::cout << "============================\n";
    }

private:

    // Action clients
    rclcpp_action::Client<FollowJointTrajectory>::SharedPtr arm_client_;
    rclcpp_action::Client<FollowJointTrajectory>::SharedPtr gripper_client_;

    // State
    std::atomic<bool> is_executing_;
    double pending_duration_;

    // Thread
    std::thread input_thread_;

    // Presets
    std::map<std::string, std::vector<double>> arm_presets_;

    double GRIPPER_OPEN_;
    double GRIPPER_CLOSED_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<RobotInterface>();

    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}